### PR TITLE
[network][noise] testing fragmented reads and writes

### PR DESCRIPTION
**Pull request by mimoo**
This commit adds two socket utilities:

* ReadOnlyTestSocket. A socket to read from only.
* ExposingTestSocket. A pair of socket that can write and read.

Both types can force fragmented reads (only read byte by byte),
as well as trailing reads (once the buffer is read, continue to read 0s).
ExposingTestSocket also allows one to record what is written on the socket.

Both types are useful for fuzzing and testing.
The commit also make use of these to add tests to verify that a noise handshake and noise streams can handle fragmented reads.

### CODE REVIEW

File: network/src/noise/handshake.rs [hunk 1]
```diff
@@ -613,4 +614,26 @@ mod test {
...
+        let (mut dialer_socket, mut listener_socket) = ExposingTestSocket::new_pair();
+
+        // fragment reads
+        dialer_socket.set_fragmented_read();
+        listener_socket.set_fragmented_read();
+
+        // get peers
+        let ((client, _client_public_key), (server, server_public_key)) = build_peers(false);
+
+        // perform the handshake
```
Comment (gregnazario) [inline network/src/noise/handshake.rs:630]: I appreciate the comments!

File: network/src/noise/stream.rs [hunk 2]
```diff
@@ -672,4 +670,47 @@ mod test {
...
+        let (client, server) = block_on(join(
+            client.upgrade_outbound(dialer_socket, server_public_key, AntiReplayTimestamps::now),
+            server.upgrade_inbound(listener_socket),
+        ));
+
+        // get session
+        let mut client = client.unwrap();
+        let (mut server, _) = server.unwrap();
+
+        //
```
Comment (gregnazario) [inline network/src/noise/stream.rs:696]: Missing a comment?  Maybe about the author of all these books :P

File: network/src/noise/stream.rs [hunk 3]
```diff
@@ -672,4 +670,47 @@ mod test {
...
+        block_on(client.write_all(b"The Name of the Wind")).unwrap();
```
Comment (gregnazario) [inline network/src/noise/stream.rs:697]: Nit: I'd normally like to see these in a variable and reused twice, especially then you don't accidentally type another space or whatever,  but the readability is nice here.

File: network/src/testutils/fake_socket.rs [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    pub fn new(content: &'a [u8]) -> Self {
+        Self {
+            content,
+            fragmented: false,
+            trailing: false,
+        }
+    }
+
+    /// reads will have to be done byte by byte
+    #[allow(dead_code)]
```
Comment (gregnazario) [inline network/src/testutils/fake_socket.rs:52]: Why are these dead code?

File: network/src/testutils/fake_socket.rs [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+                }
+                return Poll::Ready(Ok(buf.len()));
+            } else {
+                // EOF
+                return Poll::Ready(Ok(0));
+            }
+        }
+
+        // if something left return what's asked
+        if self.fragmented {
```
Comment (gregnazario) [inline network/src/testutils/fake_socket.rs:106]: So fragmented will always be exactly 1 byte?  I think it's a fine estimation, just wondering if you'd want to put some variability in it.

File: network/src/testutils/fake_socket.rs [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+            self.content = &self.content[to_read..self.content.len()];
+
+            // return length read
+            Poll::Ready(Ok(to_read))
+        }
+    }
+}
+
+//
+// ExposingTestSocket
```
Comment (gregnazario) [inline network/src/testutils/fake_socket.rs:131]: Just wondering, what does "ExposingTestSocket" mean?  it's a test socket that shows what's on the socket?

File: network/src/testutils/fake_socket.rs [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+//! What is this fake socket for?
+//!
+//! - does nothing on a write
+//! - read behaves according to mode
+//!
+//!
+//! used to quickly consume fuzzing data and dismiss socket writes.
+//! ```
+//!
+//! ```
```
Comment (phlip9) [inline network/src/testutils/fake_socket.rs:18]: cleanup?

File: network/src/testutils/fake_socket.rs [hunk 8]
```diff
@@ -0,0 +1,317 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//!
+//! This module exposes two types of sockets useful for tests:
+// - ExposingTestSocket: a wrapper around MemorySocket to retrieve handshake messages being sent.
+// - ReadOnlyTestSocket: a socket that can be read from in different ways.
```
Comment (phlip9) [inline network/src/testutils/fake_socket.rs:7]: missed "!" in these two lines (should be "//!")

File: network/src/testutils/fake_socket.rs [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        }
+    }
+
+    /// the vec passed as argument will expand to store any writes on the socket
+    pub fn save_writting(&mut self, buf: &'a mut Vec<u8>) {
+        self.written = Some(buf);
+    }
+
+    /// reads will have to be done byte by byte
+    #[allow(dead_code)]
```
Comment (phlip9) [inline network/src/testutils/fake_socket.rs:159]: do we still need this?

File: network/src/testutils/fake_socket.rs [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+}
+
+impl<'a> AsyncRead for ExposingTestSocket<'a> {
+    fn poll_read(
+        mut self: Pin<&mut Self>,
+        context: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // read from inner sockets
+        let mut buf_byte = [0u8];
```
Comment (phlip9) [inline network/src/testutils/fake_socket.rs:204]: move buf_byte into the if branch

File: network/src/testutils/fake_socket.rs (hunk 7 above)
Comment (mimoo) [reply network/src/testutils/fake_socket.rs:18]: missed that :D

File: network/src/testutils/fake_socket.rs (hunk 4 above)
Comment (mimoo) [reply network/src/testutils/fake_socket.rs:52]: ah! I figured out why I was getting linting errors. I forgot to put the testutils module under a `#[cfg(test)]`

File: network/src/testutils/fake_socket.rs (hunk 5 above)
Comment (mimoo) [reply network/src/testutils/fake_socket.rs:106]: I thought about it, and it'd be easy to put some variability in it, but I don't think it's necessary. We can always add that later if we need it : o

File: network/src/testutils/fake_socket.rs (hunk 6 above)
Comment (mimoo) [reply network/src/testutils/fake_socket.rs:131]: Better name indeed! I'll rename

File: network/src/testutils/fake_socket.rs (hunk 9 above)
Comment (mimoo) [reply network/src/testutils/fake_socket.rs:159]: got rid of it, forgot to have testutils behind a test flag!

Comment (mimoo) [reply network/src/testutils/fake_socket.rs:149]: I'm getting super weird errors that aren't connected at all to this PR oO

```
error[E0599]: no function or associated item named `mock` found for struct `libra_config::network_id::NetworkContext` in the current scope
   --> network/src/connectivity_manager/test.rs:178:93
    |
178 |             peer_manager::ConnectionNotification::NewPeer(peer_id, address, NetworkContext::mock()),
    |                                                                                             ^^^^ function or associated item not found in `libra_config::network_id::NetworkContext`
```

Comment (gregnazario) [reply network/src/testutils/fake_socket.rs:149]: I got those locally, and I found when I pulled all the updates and rebuilt it was fine, or at least on the build server it was fine.

File: network/src/noise/fuzzing.rs [hunk 11]
```diff
@@ -103,7 +51,11 @@ fn generate_first_two_messages() -> (Vec<u8>, Vec<u8>) {
     let responder = NoiseUpgrader::new(peer_id, private_key, HandshakeAuthMode::ServerOnly);
 
     // create exposing socket
-    let (dialer_socket, listener_socket) = ExposingSocket::new_pair();
+    let (mut dialer_socket, mut listener_socket) = ExposingTestSocket::new_pair();
+    let mut init_msg = Vec::new();
+    let mut resp_msg = Vec::new();
+    dialer_socket.save_writting(&mut init_msg);
```
[the remark this answers is not in the document]
Comment (mimoo) [reply network/src/noise/fuzzing.rs]: D:

File: network/src/lib.rs [hunk 12]
```diff
@@ -19,6 +19,7 @@ pub mod validator_network;
 pub mod counters;
 mod peer;
 mod sink;
+mod testutils;
```
[the remark this answers is not in the document]
Comment (mimoo) [reply network/src/lib.rs]: yep, did just that!

**Pull request closed without merging**